### DISCUSSION TITLE
Add new instance of livewire component on button click

### DISCUSSION THREAD

**Discussion by cbhai** [livewire/livewire#3398] [1. Help] [answered]
All the examples of nested livewire I see on internet use foreach loop. If loops runs 10 times, it can create 10 instances of livewire component.

Is it possible to add new instance of livewire component on button click. Let's say I have a livewire component named 'detail-card'. If there is button "Add Details" in parent page blade or different livewire component on same page. Everytime on pressing "Add Details", I want new instance of "detail-card" (livewire component) to get added at bottom of all existing "detail-card" instances.

**Comment by joshhanley** [comment, marked as answer]
@cbhai yes, just keep a counter and use a for loop instead of a for each loop.

```blade
@for($i = 0; $i < $numberOfDetailComponents; $i++)
    <livewire:detail-component ... />
@endfor
```

Hope this helps!

**Reply by cbhai** [reply]
I am not sure if this answers the question. What I am trying to find is. Page is loaded, now if you click "Add" button, a new livewire instance gets added in page above button. If I click button 10 times, on each button click, an instance of livewire component is created and added in a row fashion one after other.

**Reply by joshhanley** [reply]
@cbhai yeah it would work, you just need to increment `$numberOfDetailComponents` to add an extra one to the end.

Like
```html
<div>
    @for($i = 0; $i < $numberOfDetailComponents; $i++)
        <livewire:detail-component ... />
    @endfor
</div>

<button type="button wire:click="addDetailComponent">Add Detail</button>
```

Then in your parent component have this property and method
```php
public $numberOfDetailComponents = 1;

public function addDetailComponet() {
    $this->numberOfDetailComponents++;
}
```

This above would work, but you'd need to bind each component to some data in an array or collection, which is why other examples use a `foreach` loop. Either way using a foreach or a for loop will work fine, and is what is needed to get it to run.

There is no way to just add components in like you would do with say javascript and adding a new element to the page, as Livewire uses blade to add extra components.

Hope this helps!

**Reply by cbhai** [reply]
Yes this works. Thank you so much. Thank you for your help.

**Comment by kylewallace** [comment]
This is a stripped down example of how I am handling nested forms currently. I am using it to handle nested arrays several levels deep and it seems to suit my needs. I am in no way even close to an expert with PHP so please excuse any errors. It might give you an alternative to try.

Parent Livewire Component:

```
public $form = [];

public function addDetail()
{
    $this->form['details'][] = []; // each array item will contain the data you need to pass to each component
}

public function removeDetail($detail)
{
    unset($this->form['details'][$detail]); // remove a detail from the array
    
    $this->form['details'] = array_values($this->form['details']); // reindex the array
}
```

Parent Blade File:

```
@foreach($form['details'] as $detail)
    @livewire('detail-card', ['index' => $loop->index, 'detail' => $detail], key("detail-{$loop->index}") // you will probably want to key each component
@endforeach

<button type="button" wire:click="addDetail">Add Detail</button>
```

I opt not to use nested Livewire components if I can help it as this example manages the data structure in the parent component under the `$form` array. You would have to rely on events and listeners to keep the array up to date.

If you use the nested component you would likely have the remove button inside each component. I've added the `$index` property to the `@livewire directive` (not sure if you can just access the key). You could use that `$index` property on a button in the component to remove it from the parent array:

`<button type="button" wire:click="removeDetail($index)">Add Detail</button>`

`removeDetail` could then be a function inside the loop component which emits an event (i.e. 'detail-removed') which the parent will listen for and receive the `$index`, which is in turn passed to the `removeDetail` function on the parent component.

If you don't use nested Livewire components you don't need to worry about the events and just call it all inside the Livewire component.

Hope that makes sense.
